### PR TITLE
Fix bug in ZStream#drainFork

### LINKED ISSUE #2466: StreamDrainFork test is flaky

### PR DESCRIPTION
**Pull request by iravid**
Fixes #2466.

### CODE REVIEW

**Comment by iravid** [comment]
@vasilmkd Could you have a look?

File: streams/shared/src/main/scala/zio/stream/ZStream.scala [hunk 1]
```diff
@@ -1692,7 +1692,15 @@ class ZStream[-R, +E, +A] private[stream] (private[stream] val structure: ZStrea
         done <- Ref.make(false).toManaged_
         pull = done.get flatMap {
           if (_) Pull.end
-          else as.raceAttempt(p.await.mapError(Some(_)).ensuring(done set true) *> Pull.end)
+          else
+            as.raceAttempt(
```
Comment (regiskuckaertz) [inline streams/shared/src/main/scala/zio/stream/ZStream.scala]: Hmm, should you set `done` to `true` if `as` completes successfully?

Comment (regiskuckaertz) [reply streams/shared/src/main/scala/zio/stream/ZStream.scala]: Ah, I misread the bracketing. Looks great!

Comment (iravid) [reply streams/shared/src/main/scala/zio/stream/ZStream.scala]: Yeah we only toggle done if the interrupting promise resolves. Thanks for looking!

File: streams/shared/src/main/scala/zio/stream/ZStream.scala [hunk 2]
```diff
@@ -1692,7 +1692,15 @@ class ZStream[-R, +E, +A] private[stream] (private[stream] val structure: ZStrea
...
+              p.await
+                .mapError(Some(_))
+                .foldCauseM(
+                  c => done.set(true) *> ZIO.halt(c),
```
Comment (vasilmkd) [inline streams/shared/src/main/scala/zio/stream/ZStream.scala]: `Pull.halt`

**Pull request merged**